llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp

```diff
@@ -0,0 +1,47 @@
+//===------------- CleanupSYCLCompilerInternalMetadata.cpp --------------------
+//------------ remove metadata introduced by SYCL compiler MD -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Cleanup SYCL compiler internal metadata inserted by the frontend as it will
+// never be used in the compilation ever again
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h"
+
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace {
+
+void cleanupSYCLCompilerModuleMetadata(const Module &M, llvm::StringRef MD) {
+  NamedMDNode *Node = M.getNamedMetadata(MD);
+  if (!Node)
+    return;
+  Node->clearOperands();
```

Comment:
That only drops references to those operands, but doesn't remove operands themselves, right?

What happens to such "dead" operands later at `llvm-link` stage?

Reply:
That's a funny question, in IR I see no dead nodes and everything works just fine, while the description and implementation of clearOperands indeed just drops references. I'll double check tomorrow.

Reply:
If I read the code correctly we will end up calling d'tor for md operands:
[SmallVectorTemplateBase](https://llvm.org/doxygen/classllvm_1_1SmallVectorTemplateBase.html#a59bf49ac292d1f2d6ab5db4be8d98f3d)(size_t [Size](https://llvm.org/doxygen/classllvm_1_1SmallVectorBase.html#a7b5a03b19133c790a4d6fff66a5d2135)) : [SmallVectorTemplateCommon](https://llvm.org/doxygen/classllvm_1_1SmallVectorTemplateCommon.html)<[T](https://llvm.org/doxygen/classT.html)>([Size](https://llvm.org/doxygen/classllvm_1_1SmallVectorBase.html#a7b5a03b19133c790a4d6fff66a5d2135)) {}
 
  static void [destroy_range](https://llvm.org/doxygen/classllvm_1_1SmallVectorTemplateBase.html#a2d25f6f9379888156d4dc87f0c70a0da)([T](https://llvm.org/doxygen/classT.html) *S, [T](https://llvm.org/doxygen/classT.html) *[E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c)) {
    while (S != [E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c)) {
      --[E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c);
      [E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c)->~T();
    }
  }

so the operands are not "dead", but dead.

Reply:
Thanks!

I would suggest to add a test (could be an update to any existing one) which makes it so that one of operands of removed metadata is also an operand of a non-removed metadata. To make sure that it doesn't crash, at least. Because with simple metadata as `!N = !{i32 4}`, there is very high chance that it is used by more than one metadata node in a bigger application.